java-checks/src/main/java/org/sonar/java/checks/PathVariableAnnotationShouldBePresentIfPathVariableIsUsedCheck.java

```diff
@@ -51,18 +53,30 @@ public class PathVariableAnnotationShouldBePresentIfPathVariableIsUsedCheck exte
 
   @Override
   public List<Tree.Kind> nodesToVisit() {
-    return List.of(Tree.Kind.METHOD);
+    return List.of(Tree.Kind.CLASS);
   }
 
   @Override
   public void visitNode(Tree tree) {
-    MethodTree method = (MethodTree) tree;
+    ClassTree clazzTree = (ClassTree) tree;
 
-    MAPPING_ANNOTATIONS
-      .forEach(annotation -> reportIssueOnParameters(method, annotation));
+    List<MethodTree> methods = clazzTree.members().stream()
+      .filter(member -> member.is(Tree.Kind.METHOD))
+      .map(MethodTree.class::cast)
+      .collect(Collectors.toList());
+
+    Set<String> modelAttributePathVariable = methods.stream()
+      .filter(method -> method.symbol().metadata().isAnnotatedWith(MODEL_ATTRIBUTE_ANNOTATION))
+      .flatMap(method -> method.parameters().stream())
+      .map(variable -> pathVariableName(variable))
+      .flatMap(Optional::stream)
+      .collect(Collectors.toSet());
+
+    methods.forEach(method -> MAPPING_ANNOTATIONS
+      .forEach(annotation -> reportIssueOnParameters(method, annotation, modelAttributePathVariable)));
   }
 
-  private void reportIssueOnParameters(MethodTree method, String annotation) {
+  private void reportIssueOnParameters(MethodTree method, String annotation, Set<String> modelAttributePathVariable) {
```

Comment:
This method does too many things at once. I would split it into two or three methods. Also about the naming: I think we have the implicit convention that a method should be called `reportSomething` only when it _definitely_ reports something (like: all paths in the method lead to it). If it may report or not, we should call it `checkSomething` (or something similar that fits the purpose of the method).

About splitting: I would split it into `containsMapParameter`, sth. like `findMissingPathVariables` and then the actual reporting:

```java
private void checkParameters(...) {
  if (!containsMapParameter) {
    return;
  }
  findMissingPathVariables(...).forEach(MissingVariableAnnotationCheck::reportMissingPathVariable);
}
```